File: android/lib/src/main/kotlin/com/artifex/mupdf/mobile/MuPDFDocument.kt
package com.artifex.mupdf.mobile

import android.util.SizeF
import java.io.File

/**
 * A PDF (or XPS / CBZ / EPUB) document managed by the MuPDF engine.
 *
 * ## Thread safety
 * `MuPDFDocument` is **not** thread-safe. Serialise all calls externally when
 * using multiple threads.
 *
 * ## Memory management
 * Always call [close] (or use Kotlin's `use` block) when finished. The JVM
 * finaliser also closes the document, but relying on GC for native resource
 * release is discouraged.
 *
 * ## Example
 * ```kotlin
 * MuPDFDocument.open("/sdcard/sample.pdf").use { doc ->
 *     println("Pages: ${doc.pageCount}")
 *     val page = doc.loadPage(0)
 *     val bitmap = page.render(2f)
 *     imageView.setImageBitmap(bitmap.bitmap)
 * }
 * ```
 */
open class MuPDFDocument private constructor(
    /** Opaque pointer to the native `DocHandle`. */
    internal var nativeHandle: Long
) : AutoCloseable {

    private var closed = false

    // ─────────────────────────────────────────────────────────────────────────
    // Companion / factory
    // ─────────────────────────────────────────────────────────────────────────

    companion object {
        init {
            System.loadLibrary("mupdf_mobile")
        }

        /**
         * Opens a document at the given file-system path.
         *
         * @param path Absolute path to the PDF (or other supported format).
         * @throws MuPDFFileNotFoundException if the path does not exist.
         * @throws MuPDFInvalidDocumentException if the file cannot be parsed.
         */
        @JvmStatic
        @Throws(MuPDFFileNotFoundException::class, MuPDFInvalidDocumentException::class)
        fun open(path: String): MuPDFDocument {
            if (!File(path).exists()) throw MuPDFFileNotFoundException(path)
            val handle = nativeOpen(path)
            if (handle == -1L) throw MuPDFInvalidDocumentException()
            return MuPDFDocument(handle)
        }

        /**
         * Opens a document from in-memory bytes.
         *
         * @param data Raw bytes of a supported document format.
         * @throws MuPDFInvalidDocumentException if the data cannot be parsed.
         */
        @JvmStatic
        @Throws(MuPDFInvalidDocumentException::class)
        fun open(data: ByteArray): MuPDFDocument {
            if (data.isEmpty()) throw MuPDFInvalidDocumentException()
            val handle = nativeOpenFromBytes(data)
            if (handle == -1L) throw MuPDFInvalidDocumentException()
            return MuPDFDocument(handle)
        }

        // JNI bridge
        @JvmStatic private external fun nativeOpen(path: String): Long
        @JvmStatic private external fun nativeOpenFromBytes(data: ByteArray): Long
        @JvmStatic private external fun nativeGetPageCount(docHandle: Long): Int
        @JvmStatic private external fun nativeLoadPage(docHandle: Long, index: Int): Long
        @JvmStatic private external fun nativeGetPageWidth(docHandle: Long, index: Int): Float
        @JvmStatic private external fun nativeGetPageHeight(docHandle: Long, index: Int): Float
        @JvmStatic private external fun nativeSave(docHandle: Long, path: String): Boolean
        @JvmStatic private external fun nativeMerge(dstHandle: Long, srcHandle: Long)
        @JvmStatic private external fun nativeInsertBlankPage(docHandle: Long, index: Int, width: Float, height: Float)
        @JvmStatic private external fun nativeDeletePage(docHandle: Long, index: Int)
        @JvmStatic private external fun nativeGetMetadata(docHandle: Long, key: String): String?
        @JvmStatic private external fun nativeClose(docHandle: Long)
        @JvmStatic private external fun nativeGetOutline(docHandle: Long): Array<String>
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Properties
    // ─────────────────────────────────────────────────────────────────────────

    /** Total number of pages in the document. */
    val pageCount: Int
        get() {
            checkNotClosed()
            if (nativeHandle == -1L) return 0
            return nativeGetPageCount(nativeHandle)
        }

    /** The document's title metadata, or `null` if absent. */
    val title: String?
        get() {
            checkNotClosed()
            if (nativeHandle == -1L) return null
            return nativeGetMetadata(nativeHandle, "info:Title")
        }

    /** The document's author metadata, or `null` if absent. */
    val author: String?
        get() {
            checkNotClosed()
            if (nativeHandle == -1L) return null
            return nativeGetMetadata(nativeHandle, "info:Author")
        }

    // ─────────────────────────────────────────────────────────────────────────
    // Page access
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Loads the page at the given zero-based index.
     *
     * @param index Zero-based page index (0 until [pageCount]).
     * @throws MuPDFDocumentClosedException if the document is closed.
     * @throws MuPDFPageOutOfBoundsException if [index] is out of range.
     */
    @Throws(MuPDFDocumentClosedException::class, MuPDFPageOutOfBoundsException::class)
    open fun loadPage(index: Int): MuPDFPage {
        checkNotClosed()
        val count = pageCount
        if (index < 0 || index >= count) {
            throw MuPDFPageOutOfBoundsException(index, count)
        }
        val pageHandle = nativeLoadPage(nativeHandle, index)
        if (pageHandle == -1L) throw MuPDFEngineException("Failed to load page $index")
        val w = nativeGetPageWidth(nativeHandle, index)
        val h = nativeGetPageHeight(nativeHandle, index)
        return MuPDFPage(index, w, h, pageHandle)
    }

    /**
     * Returns the size (width × height in PDF points) of the page at [index]
     * without loading the full page object.
     */
    open fun pageSize(at: Int): SizeF {
        checkNotClosed()
        if (nativeHandle == -1L || at < 0 || at >= pageCount) return SizeF(0f, 0f)
        val w = nativeGetPageWidth(nativeHandle, at)
        val h = nativeGetPageHeight(nativeHandle, at)
        return SizeF(w, h)
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Save
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Saves the document (with any modifications) to [path].
     *
     * @throws MuPDFDocumentClosedException if the document is closed.
     * @throws MuPDFSaveException if the write fails.
     */
    @Throws(MuPDFDocumentClosedException::class, MuPDFSaveException::class)
    open fun save(path: String) {
        checkNotClosed()
        if (nativeHandle == -1L) return
        val ok = nativeSave(nativeHandle, path)
        if (!ok) throw MuPDFSaveException(path)
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Editing
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Appends all pages from [document] to the end of this document.
     *
     * @throws MuPDFDocumentClosedException if either document is closed.
     */
    @Throws(MuPDFDocumentClosedException::class)
    open fun merge(document: MuPDFDocument) {
        checkNotClosed()
        if (document.closed) throw MuPDFDocumentClosedException()
        if (nativeHandle == -1L || document.nativeHandle == -1L) return
        nativeMerge(nativeHandle, document.nativeHandle)
    }

    /**
     * Inserts a blank page at [index] with the given dimensions.
     *
     * @param index  Zero-based position for the new page.
     * @param width  Page width in PDF points.
     * @param height Page height in PDF points.
     * @throws MuPDFDocumentClosedException if the document is closed.
     */
    @Throws(MuPDFDocumentClosedException::class)
    open fun insertBlankPage(index: Int, width: Float, height: Float) {
        checkNotClosed()
        if (nativeHandle == -1L) return
        nativeInsertBlankPage(nativeHandle, index, width, height)
    }

    /**
     * Deletes pages at the specified zero-based indices.
     *
     * Indices are processed in descending order to avoid shifting.
     *
     * @throws MuPDFDocumentClosedException if the document is closed.
     * @throws MuPDFPageOutOfBoundsException if any index is out of range.
     */
    @Throws(MuPDFDocumentClosedException::class, MuPDFPageOutOfBoundsException::class)
    open fun deletePages(indices: List<Int>) {
        checkNotClosed()
        val count = pageCount
        indices.sortedDescending().forEach { index ->
            if (index < 0 || index >= count) throw MuPDFPageOutOfBoundsException(index, count)
            if (nativeHandle != -1L) nativeDeletePage(nativeHandle, index)
        }
    }

    /**
     * Returns the document's table of contents as a list of [MuPDFOutlineItem] roots.
     * Returns an empty list if the document has no outline or is closed.
     */
    open fun outline(): List<MuPDFOutlineItem> {
        checkNotClosed()
        if (nativeHandle == -1L) return emptyList()
        val flat = nativeGetOutline(nativeHandle)
        return parseOutlineItems(flat)
    }

    private fun parseOutlineItems(flat: Array<String>): List<MuPDFOutlineItem> {
        if (flat.isEmpty()) return emptyList()
        // Stack-based DFS reconstruction from the flat depth-annotated list.
        data class Entry(val depth: Int, val item: MuPDFOutlineItem, val children: MutableList<MuPDFOutlineItem>)
        val stack = mutableListOf<Entry>()
        val roots = mutableListOf<MuPDFOutlineItem>()

        for (line in flat) {
            val parts = line.split("\t", limit = 3)
            if (parts.size < 3) continue
            val depth = parts[0].toIntOrNull() ?: 0
            val title = parts[1]
            val pageIndex = parts[2].toIntOrNull() ?: -1
            val children = mutableListOf<MuPDFOutlineItem>()
            val item = MuPDFOutlineItem(title, pageIndex, children)
            val entry = Entry(depth, item, children)

            // Pop stack back to the parent level.
            while (stack.isNotEmpty() && stack.last().depth >= depth) {
                stack.removeLast()
            }
            if (stack.isEmpty()) {
                roots.add(item)
            } else {
                stack.last().children.add(item)
            }
            stack.add(entry)
        }
        return roots
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Lifecycle
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Releases all native resources. Safe to call multiple times.
     *
     * After calling `close()`, further operations will throw
     * [MuPDFDocumentClosedException].
     */
    override fun close() {
        if (closed) return
        closed = true
        if (nativeHandle != -1L) {
            nativeClose(nativeHandle)
            nativeHandle = -1L
        }
    }

    @Suppress("ProtectedInFinal")
    protected fun finalize() {
        close()
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Internal helpers
    // ─────────────────────────────────────────────────────────────────────────

    private fun checkNotClosed() {
        if (closed) throw MuPDFDocumentClosedException()
    }
}

File: Sources/MuPDFMobile/MuPDFDocument.swift
import Foundation
#if canImport(CoreGraphics)
import CoreGraphics
#endif

// MARK: - MuPDFDocument

/// A PDF (or XPS / CBZ / EPUB) document managed by the MuPDF engine.
///
/// ## Thread safety
/// Instances of `MuPDFDocument` are **not** thread-safe. All calls to a given
/// document must be made from the same thread (or serialised externally).
/// Use separate documents per background thread if concurrent rendering is
/// required.
///
/// ## Memory management
/// Call ``close()`` to release native resources as soon as the document is no
/// longer needed. The ARC destructor also calls `close()` as a safety net.
#if canImport(ObjectiveC)
@objcMembers
#endif
public final class MuPDFDocument: NSObject {

    // -------------------------------------------------------------------------
    // MARK: Private state
    // -------------------------------------------------------------------------

    /// Opaque pointer to the underlying `fz_document`.
    private var nativeDoc: OpaquePointer?

    /// Whether the document has been explicitly closed.
    private var isClosed = false

    // -------------------------------------------------------------------------
    // MARK: Initialisation (private — use factory methods)
    // -------------------------------------------------------------------------

    private override init() {}

    private init(nativeDoc: OpaquePointer?) {
        self.nativeDoc = nativeDoc
    }

    deinit {
        close()
    }

    // -------------------------------------------------------------------------
    // MARK: Factory – open from path
    // -------------------------------------------------------------------------

    /// Opens a document at the given file-system path.
    ///
    /// - Parameter path: Absolute path to the PDF (or other supported format).
    /// - Returns: An initialised `MuPDFDocument`.
    /// - Throws: `MuPDFError.fileNotFound` if the path does not exist.
    ///           `MuPDFError.invalidDocument` if the file cannot be parsed.
    public static func open(path: String) throws -> MuPDFDocument {
        guard FileManager.default.fileExists(atPath: path) else {
            throw MuPDFError.fileNotFound
        }

        // TODO: (requires mupdf submodule)
        //   let ctx = MuPDFContext.shared.ctx
        //   guard let doc = fz_open_document(ctx, path) else {
        //       throw MuPDFError.invalidDocument
        //   }
        //   return MuPDFDocument(nativeDoc: doc)

        // Placeholder until submodule is available:
        let doc = MuPDFDocument()
        doc.nativeDoc = nil
        return doc
    }

    /// Opens a document from in-memory data.
    ///
    /// - Parameter data: The raw bytes of a PDF (or other supported format).
    /// - Returns: An initialised `MuPDFDocument`.
    /// - Throws: `MuPDFError.invalidDocument` if the data cannot be parsed.
    public static func open(data: Data) throws -> MuPDFDocument {
        guard !data.isEmpty else { throw MuPDFError.invalidDocument }

        // TODO: (requires mupdf submodule)
        //   data.withUnsafeBytes { ptr in
        //       let buf = fz_new_buffer_from_copied_data(ctx, ptr.baseAddress, data.count)
        //       let stream = fz_open_buffer(ctx, buf)
        //       let doc = fz_open_document_with_stream(ctx, "application/pdf", stream)
        //       ...
        //   }

        let doc = MuPDFDocument()
        doc.nativeDoc = nil
        return doc
    }

    // -------------------------------------------------------------------------
    // MARK: Document properties
    // -------------------------------------------------------------------------

    /// The total number of pages in the document.
    public var pageCount: Int {
        guard !isClosed else { return 0 }
        // TODO: return Int(fz_count_pages(MuPDFContext.shared.ctx, nativeDoc))
        return 0
    }

    /// The document's title metadata, if present.
    public var title: String? {
        guard !isClosed else { return nil }
        // TODO: return metadata(key: "info:Title")
        return nil
    }

    /// The document's author metadata, if present.
    public var author: String? {
        guard !isClosed else { return nil }
        // TODO: return metadata(key: "info:Author")
        return nil
    }

    // -------------------------------------------------------------------------
    // MARK: Page access
    // -------------------------------------------------------------------------

    /// Loads the page at the given zero-based index.
    ///
    /// - Parameter index: Zero-based page index (0 …< ``pageCount``).
    /// - Returns: The loaded ``MuPDFPage``.
    /// - Throws: `MuPDFError.documentClosed` or
    ///           `MuPDFError.pageIndexOutOfBounds`.
    public func loadPage(at index: Int) throws -> MuPDFPage {
        guard !isClosed else { throw MuPDFError.documentClosed }
        guard index >= 0 && index < pageCount else {
            throw MuPDFError.pageIndexOutOfBounds
        }

        // TODO: (requires mupdf submodule)
        //   let ctx = MuPDFContext.shared.ctx
        //   guard let nativePage = fz_load_page(ctx, nativeDoc, Int32(index)) else {
        //       throw MuPDFError.mupdfError
        //   }
        //   return MuPDFPage(nativePage: nativePage, index: index, document: self)

        return MuPDFPage(index: index, document: self)
    }

    /// Returns the size (in PDF points) of the page at the given index
    /// without loading the full page object.
    ///
    /// - Parameter index: Zero-based page index.
    /// - Returns: The page size as a `CGSize`.
#if canImport(CoreGraphics)
    public func pageSize(at index: Int) -> CGSize {
        guard !isClosed, index >= 0, index < pageCount else { return .zero }
        // TODO: use fz_bound_page after loading
        return .zero
    }
#endif

    // -------------------------------------------------------------------------
    // MARK: Save
    // -------------------------------------------------------------------------

    /// Saves the document (with any modifications) to the specified path.
    ///
    /// - Parameter path: The destination file path.
    /// - Throws: `MuPDFError.documentClosed` or `MuPDFError.saveFailed`.
    public func save(to path: String) throws {
        guard !isClosed else { throw MuPDFError.documentClosed }
        guard !path.isEmpty else { throw MuPDFError.saveFailed }

        // TODO: (requires mupdf submodule)
        //   let ctx = MuPDFContext.shared.ctx
        //   let pdfDoc = pdf_document_from_fz_document(ctx, nativeDoc)
        //   pdf_save_document(ctx, pdfDoc, path, nil)
    }

    // -------------------------------------------------------------------------
    // MARK: Document editing
    // -------------------------------------------------------------------------

    /// Appends all pages from `document` to the end of this document.
    ///
    /// - Parameter document: The source document to merge from.
    /// - Throws: `MuPDFError.documentClosed` or `MuPDFError.mupdfError`.
    public func merge(document: MuPDFDocument) throws {
        guard !isClosed else { throw MuPDFError.documentClosed }
        guard !document.isClosed else { throw MuPDFError.documentClosed }

        // TODO: (requires mupdf submodule)
        //   let ctx = MuPDFContext.shared.ctx
        //   let dst = pdf_document_from_fz_document(ctx, nativeDoc)
        //   let src = pdf_document_from_fz_document(ctx, document.nativeDoc)
        //   pdf_merge_document(ctx, dst, src)
    }

    /// Inserts a blank page with the given dimensions at the specified index.
    ///
    /// - Parameters:
    ///   - index: Zero-based position at which to insert the new page.
    ///   - width:  Page width in PDF points (1 point = 1/72 inch).
    ///   - height: Page height in PDF points.
    /// - Throws: `MuPDFError.documentClosed` or `MuPDFError.mupdfError`.
    public func insertBlankPage(at index: Int, width: Float, height: Float) throws {
        guard !isClosed else { throw MuPDFError.documentClosed }

        // TODO: (requires mupdf submodule)
        //   let ctx = MuPDFContext.shared.ctx
        //   let pdfDoc = pdf_document_from_fz_document(ctx, nativeDoc)
        //   let mediabox = fz_make_rect(0, 0, width, height)
        //   pdf_insert_page(ctx, pdfDoc, Int32(index), pdf_add_page(ctx, pdfDoc, mediabox, 0, nil, nil))
    }

    /// Deletes pages at the specified zero-based indices.
    ///
    /// Indices are sorted and processed in reverse order so earlier deletions
    /// do not shift subsequent indices.
    ///
    /// - Parameter indices: The set of page indices to delete.
    /// - Throws: `MuPDFError.documentClosed`, `MuPDFError.pageIndexOutOfBounds`,
    ///           or `MuPDFError.mupdfError`.
    public func deletePages(at indices: [Int]) throws {
        guard !isClosed else { throw MuPDFError.documentClosed }
        let sorted = indices.sorted(by: >)
        for index in sorted {
            guard index >= 0 && index < pageCount else {
                throw MuPDFError.pageIndexOutOfBounds
            }
            // TODO: (requires mupdf submodule)
            //   let ctx = MuPDFContext.shared.ctx
            //   let pdfDoc = pdf_document_from_fz_document(ctx, nativeDoc)
            //   pdf_delete_page(ctx, pdfDoc, Int32(index))
        }
    }

    // -------------------------------------------------------------------------
    // MARK: Close / release
    // -------------------------------------------------------------------------

    /// Releases all native resources held by this document.
    ///
    /// After calling `close()`, most operations will throw
    /// `MuPDFError.documentClosed`. It is safe to call `close()` multiple
    /// times.
    public func close() {
        guard !isClosed else { return }
        isClosed = true
        // TODO: (requires mupdf submodule)
        //   fz_drop_document(MuPDFContext.shared.ctx, nativeDoc)
        nativeDoc = nil
    }

    // -------------------------------------------------------------------------
    // MARK: Private helpers
    // -------------------------------------------------------------------------

    private func metadata(key: String) -> String? {
        // TODO: (requires mupdf submodule)
        //   var buf = [CChar](repeating: 0, count: 256)
        //   fz_lookup_metadata(ctx, nativeDoc, key, &buf, Int32(buf.count))
        //   return String(cString: buf).isEmpty ? nil : String(cString: buf)
        return nil
    }
}
